pkg/backoff/backoff.go

```diff
@@ -0,0 +1,123 @@
+package backoff
+
+import (
+	"fmt"
+	"math"
+	"strconv"
+	"time"
+
+	ini "github.com/wakatime/wakatime-cli/pkg/config"
+	"github.com/wakatime/wakatime-cli/pkg/heartbeat"
+	"github.com/wakatime/wakatime-cli/pkg/log"
+
+	"github.com/spf13/viper"
+)
+
+const (
+	// resetAfter sets the total seconds a backoff will last.
+	resetAfter = 3600
+	// factor is the total seconds to be multiplied by.
+	factor = 15
+)
+
+// RetryConfig defines backoff data.
+type RetryConfig struct {
+	// At is the time when the first failure happened.
+	At *time.Time
+	// Retries is the number of attempts to connect.
+	Retries int
+	// V is an instance of Viper.
+	V *viper.Viper
+}
+
+// WithRetry initializes and returns a heartbeat handle option, which
+// can be used in a heartbeat processing pipeline to prevent trying to send
+// a heartbeat when the api is unresponsive.
+func WithRetry(config RetryConfig) heartbeat.HandleOption {
```

Comment:
Great! Very good idea to include it as a pipeline handle option :+1:

Reply:
Learned with you :)